server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -33,59 +51,142 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
+        case ParsePartBody          => parsePartBody(currentByte)
+      }
+    }
+
+    private def parseInitialBoundary(currentByte: Int): Unit = {
+      val foundBoundary = initialBoundaryMatcher.matchByte(currentByte.toByte)
+      if (foundBoundary) {
+        changeState(ParsePartHeaders)
+      }
+    }
+
+    private def parsePartHeaders(): Unit = {
+      if (buffer.endsWith(CRLF)) {
+        buffer.view.map(_.toChar).mkString match {
+          case headerLine if !headerLine.isBlank => addHeader(headerLine)
+          case _                                 => changeState(ParsePartBody)
+        }
+        buffer.clear()
+      } else if (buffer.length >= bufferSize) {
+        throw new RuntimeException(
+          s"Reached max size of $bufferSize bytes before reaching newline when parsing header."
         )
-      case None => changeState(AfterBoundary)
+      }
     }
-  }
 
-  def parseMultipartBody(httpExchange: HttpExchange, boundary: String): Seq[ParsedMultiPart] = {
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
-    Iterator
-      .continually(reader.readLine())
-      .takeWhile(_ != null)
-      .foldLeft(initialParseState) { case (state, line) =>
-        state.parseState match {
-          case Default if line.startsWith(boundary)           => state.changeState(AfterBoundary)
-          case Default                                        => state
-          case AfterBoundary if line.trim.isEmpty             => state.changeState(AfterHeaderSpace)
-          case AfterBoundary                                  => state.addHeader(line)
-          case AfterHeaderSpace if !line.startsWith(boundary) => state.addBody(line.getBytes())
-          case AfterHeaderSpace                               => state.completePart()
+    private def parsePartBody(currentByte: Int): Unit = {
+      bodySize += 1
+      val foundFinalBoundary = endMatcher.matchByte(currentByte.toByte)
+      val foundBoundary = bodyBoundaryMatcher.matchByte(currentByte.toByte) || foundFinalBoundary
+      convertStreamToFileIfThresholdMet()
+
+      if (foundFinalBoundary) done = true
+
+      if (foundBoundary) {
+        val minus = if (foundFinalBoundary) endMatcher.getDelimiter.length else bodyBoundaryMatcher.getDelimiter.length
+
+        val bytesToWrite = buffer.view.slice(0, buffer.length - minus)
+        bytesToWrite.foreach(byte => stream.underlying.write(byte.toInt))
+        val bodyInputStream = stream match {
+          case FileStream(tempFile, stream) => stream.close(); new FileInputStream(tempFile)
+          case ByteStream(stream)           => new ByteArrayInputStream(stream.toByteArray)
         }
+        completePart(bodyInputStream)
+        stream = ByteStream()
+        bodySize = 0
+      } else if (bodyBoundaryMatcher.getMatches == 0 && endMatcher.getMatches == 0) {
```

Comment:
isn't it possible that we'll overflow the buffer, if we constantly get a match? e.g. the body is only-As and the boundary starts with an A?

Reply:
You're right, that could make us drop a few bytes.
I'll try to avoid that :smile: